go_lib/deckhouse-config/config-map-section.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2022 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deckhouse_config
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	kcm "github.com/flant/addon-operator/pkg/kube_config_manager"
+	"github.com/flant/addon-operator/pkg/utils"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+
+	"github.com/deckhouse/deckhouse/go_lib/deckhouse-config/conversion"
+	d8config_v1 "github.com/deckhouse/deckhouse/go_lib/deckhouse-config/v1"
+)
+
+// configMapSection is a holder for one module section from the ConfigMap.
+type configMapSection struct {
+	name        string
+	valuesKey   string
+	values      utils.Values
+	enabledFlag *bool
+}
+
+// getValuesMap returns a values map or nil for module or global section.
+func (s *configMapSection) getValuesMap() (map[string]interface{}, error) {
+	untypedValues := s.values[s.valuesKey]
+
+	isValidType := true
+	switch v := untypedValues.(type) {
+	case map[string]interface{}:
+		// Module section is not empty, and it is a map.
+		return v, nil
+	case nil:
+		// Module values are nil when ConfigMap has Enabled flag without module section.
+		// Transform empty values to the empty 'configValues' field.
+		return nil, nil
+	case string:
+		// Transform empty string to the empty 'configValues' field.
+		if v != "" {
+			isValidType = false
+		}
+	case []interface{}:
+		// Array is not a valid module section, but it is ok if array is empty, just ignore it.
+		if len(v) != 0 {
+			isValidType = false
+		}
+	default:
+		// Consider other types are not valid.
+		isValidType = false
+	}
+	if !isValidType {
+		return nil, fmt.Errorf("configmap section '%s' is not an object, need map[string]interface{}, got %T:(%+v)", s.valuesKey, untypedValues, untypedValues)
+	}
+	return nil, nil
+}
+
+// convertValues runs conversions on section values.
+// Assume that values are from cm/deckhouse, so start conversions from the version 1.
+func (s *configMapSection) convertValues() (int, map[string]interface{}, error) {
+	// TODO handle empty values.
```

Comment:
Is it for future or you just forgot about it?

Reply:
Oh, this TODO is about values becomes empty after conversion. This case is handled by callers, e.g. getModuleConfig.
I'll remove this TODO.